Persist DB tables for EE features

jhoward1994 opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the tests
- Create or update the documentation at https://github.com/strapi/documentation
- Refer to the issue you are closing in the PR description: Fix #issue
- Specify if the PR is ready to be merged or work in progress (by opening a draft PR)

Please ensure you read the Contributing Guide: https://github.com/strapi/strapi/blob/main/CONTRIBUTING.md

### How to test it?

Provide information about the environment and the path to verify the behaviour.
-->

### What does it do?

- [x] EE - Adds feature table names to the DB in order to persist them when moving between EE & CE
- [x] DB - Accounts for table names from the DB when diffing database schemas 

### Why is it needed?

Audit logs and Review Workflows have the need to persist feature settings while the user switches between CE and EE

This way once each feature is enabled once, it's tables will be saved to the DB

### Questions

- [x] Should this be opt in behaviour?

Boegie19 commented:
I would expect you to use the core store for this
if it are settings 

any specific reason you give it its own table?

Marc-Roig commented on packages/core/admin/ee/server/bootstrap.js:
```diff
@@ -2,21 +2,50 @@
+  createReservedTable,
+  findTablesThatStartWithPrefix,
+  doesReservedTableEntryExist,
+  createReservedTableEntries,
+} = require('./utils/reserved-tables');
+
+module.exports = async ({ strapi }) => {
+  const connection = strapi.db.getSchemaConnection();
+  await createReservedTable(connection);
+
+  const tablesToPersist = [];
 
-module.exports = async () => {
   const { actionProvider } = getService('permission');
 
   if (features.isEnabled('sso')) {
     await actionProvider.registerMany(actions.sso);
   }
 
   if (features.isEnabled('audit-logs')) {
+    const auditLogTables = await findTablesThatStartWithPrefix('strapi_audit_logs');
+    tablesToPersist.push(...auditLogTables);
+
     await actionProvider.registerMany(actions.auditLogs);
   }
 
+  if (features.isEnabled('review-workflows')) {
+    const reviewWorkflowTables = await findTablesThatStartWithPrefix('strapi_workflows');
+    tablesToPersist.push(...reviewWorkflowTables);
+  }
+
+  let recordsToInsert = await mapAsync(tablesToPersist, (tableName) =>
+    doesReservedTableEntryExist(tableName)
+  );
+  recordsToInsert = recordsToInsert.filter((record) => record);
+
+  if (recordsToInsert.length > 0) {
+    await createReservedTableEntries(recordsToInsert);
+  }
+
```
Good work with this :) I didn't expect this PR to be open so soon.

Was thinking of other ways we could do this, wdyt of using the core store as @Boegie19 proposed?

Marc-Roig commented on packages/core/admin/ee/server/bootstrap.js:
```diff
@@ -2,21 +2,50 @@
+} = require('./utils/reserved-tables');
+
+module.exports = async ({ strapi }) => {
+  const connection = strapi.db.getSchemaConnection();
+  await createReservedTable(connection);
```
If we used the strapi core store this part could be omitted.

Marc-Roig commented on packages/core/admin/ee/server/bootstrap.js:
```diff
@@ -2,21 +2,50 @@
+    tablesToPersist.push(...reviewWorkflowTables);
+  }
+
+  let recordsToInsert = await mapAsync(tablesToPersist, (tableName) =>
+    doesReservedTableEntryExist(tableName)
+  );
+  recordsToInsert = recordsToInsert.filter((record) => record);
+
+  if (recordsToInsert.length > 0) {
+    await createReservedTableEntries(recordsToInsert);
+  }
```
I feel the first part should be inside the createReservedTableEntries, it seems checking if the table exist should be part of the createReservedTableEntries

Boegie19 commented:
If you want to continue this route what I also think is resalable since it is easier to delete the data then we mind need to look into fixing 

https://github.com/strapi/strapi/issues/16092

Marc-Roig commented on packages/core/admin/ee/server/bootstrap.js:
```diff
@@ -2,21 +2,50 @@
     await actionProvider.registerMany(actions.auditLogs);
   }
 
+  if (features.isEnabled('review-workflows')) {
+    const reviewWorkflowTables = await findTablesThatStartWithPrefix('strapi_workflows');
+    tablesToPersist.push(...reviewWorkflowTables);
+  }
```
what if we did:
```suggestion
  if (features.isEnabled('review-workflows')) {
    await reserveTablesWithPrefix('strapi_workflows');
  }
```

We could have `reserveTablesWithPrefix` and `reserveTable` , and it could be something like:
```
const reserveTable = async (tableName) = {
  const reservedTables = await getReservedTables();
  
  // TODO Make sure it's not in the array first
  reservedTables.push(tableName)
  
    await strapi.store.set({
    type: 'core',
    key: 'reserved_tables',
    value: reservedTables,
  });
}
```

Side comment, do we care if the table does not exist 🤔 ?

Marc-Roig commented:
@Boegie19 > If you want to continue this route what I also think is resalable since it is easier to delete the data then we mind need to look into fixing
https://github.com/strapi/strapi/issues/16092

Good point, I think this is something that we should check. Still this PR is about persisting the data of a set of table names even if this EE features are disabled. 

But this made me think, maybe we could find an alternative name for 'reserved' , it could create confusion. Maybe persistedTableNames? wdyt @jhoward1994 ?

Boegie19 commented:
Fair but this pr would make that issue worse

Marc-Roig commented:
> Fair but this pr would make that issue worse

why  do you think so? we are not creating new tables in this pr, those will already exist if you are using EE.

Boegie19 commented:
> > Fair but this pr would make that issue worse
> 
> why do you think so? we are not creating new tables in this pr, those will already exist if you are using EE.

Did not realize that they already existed since they are workflow related and I thought that was not yet released.

jhoward1994 commented:
@Marc-Roig & @Boegie19 thank you for your help. I've addressed the feedback in [`7e4cea8` (#16084)](https://github.com/strapi/strapi/pull/16084/commits/7e4cea8a2410a97d3c6c7ba02af2b96601afd3d0)

nathan-pichon commented on packages/core/admin/ee/server/bootstrap.js:
```diff
@@ -14,9 +15,15 @@ module.exports = async () => {
     await actionProvider.registerMany(actions.auditLogs);
   }
 
+  if (features.isEnabled('review-workflows')) {
+    await reserveTablesWithPrefix('strapi_workflows');
+  }
+
```
Might not need that in main for now. 🤔 
What do you think of adding those lines in review workflow PR instead when this current PR will be merged in main?

jhoward1994 replied:
Good point. Okay I will revert this change and make a TODO in the RW feature PR

Marc-Roig commented on packages/core/admin/ee/server/utils/reserved-tables.js:
```diff
@@ -0,0 +1,49 @@
+ * core store
+ * @param {string} tableNamePrefix
+ */
+const reserveTablesWithPrefix = async (tableNamePrefix) => {
+  const reservedTables = (await getReservedTables()) || [];
+
+  const tableNames = await findTablesThatStartWithPrefix(tableNamePrefix);
+
+  if (!tableNames.length) {
+    return;
+  }
+
+  reservedTables.push(...tableNames.filter((name) => !reservedTables.includes(name)));
+
+  await strapi.store.set({
+    type: 'core',
+    key: 'reserved_tables',
+    value: reservedTables,
+  });
+};
```
I think this could save some start up time in case all the tables are already reserved.

```suggestion
const reserveTablesWithPrefix = async (tableNamePrefix) => {
  const reservedTables = (await getReservedTables()) || [];

  const tableNames = await findTablesThatStartWithPrefix(tableNamePrefix);
	const notReservedTableNames = tableNames.filter((name) => !reservedTables.includes(name))
	
  if (!notReservedTableNames.length) {
    return;
  }

  reservedTables.push(...notReservedTableNames);

  await strapi.store.set({
    type: 'core',
    key: 'reserved_tables',
    value: reservedTables,
  });
};
```

Marc-Roig commented on packages/core/admin/ee/server/utils/persisted-tables.js:
```diff
@@ -0,0 +1,49 @@
+ * Get all reserved table names from the core store
+ * @returns {Array}
+ */
+const getPersistedTables = async () =>
+  strapi.store.get({
+    type: 'core',
+    key: 'persisted_tables',
+  }) ?? [];
```
I think the ?? [] won't work here as the store.get promise has to be awaited first

jhoward1994 replied:
ah yes you're right

nathan-pichon commented on packages/core/admin/ee/server/utils/persisted-tables.js:
```diff
@@ -0,0 +1,49 @@
+ */
+
+const persistTablesWithPrefix = async (tableNamePrefix) => {
+  const persistedTables = (await getPersistedTables()) || [];
```
`|| []` looks like overkill as getPersistedTables already returns an empty array, no?

Marc-Roig reviewed (approved):
I have added @remidej as a reviewer too to take a look, as we are persisting audit log tables :)

remidej reviewed (approved):
The audit logs part of the PR looks good to me, thanks for managing this issue :)
